sql-plugin/src/main/java/com/nvidia/spark/rapids/JCudfUtil.java

```diff
@@ -0,0 +1,586 @@
+  }
+
+  /**
+   * A method used to order the columns to pack the schema based on the JCudf specifications.
+   * It sorts fixed-width columns by their length.
+   * Variable-length columns store 8 bytes, but are 4 byte aligned since they are two 4-byte
+   * values.
+   * However, if both datatypes are variable-length, then we want the type with larger alignment to
+   * come last for better packing of the data.
+   *
+   * @param dt1 first datatype to compare.
+   * @param dt2 second datatype to compare.
+   * @return true when dt1 has more precedence than dt2.
+   */
+  public static boolean compareDataTypePrecedence(DataType dt1, DataType dt2) {
+    int refSize1;
+    int refSize2;
+    int dataAlignSize1 = 0;
+    int dataAlignSize2 = 0;
+    if (isFixedLength(dt1)){
+      refSize1 = getSizeForFixedLengthDataType(dt1);
+    } else {
+      refSize1 = JCUDF_VAR_LENGTH_FIELD_ALIGNMENT;
+      dataAlignSize1 = getDataAlignmentForDataType(GpuColumnVector.getNonNestedRapidsType(dt1));
+    }
+    if (isFixedLength(dt2)){
+      refSize2 = getSizeForFixedLengthDataType(dt2);
+    } else {
+      refSize2 = JCUDF_VAR_LENGTH_FIELD_ALIGNMENT;
+      dataAlignSize2 = getDataAlignmentForDataType(GpuColumnVector.getNonNestedRapidsType(dt2));
+    }
+    if (dataAlignSize2 > 0 && dataAlignSize1 > 0) { // both types are variable length
+      return dataAlignSize1 < dataAlignSize2;
+    }
+    return refSize1 > refSize2;
+  }
+
+  /**
+   * Sets the alignment for variable length data types.
+   * This is used to align the data in the variable width section of the JCUDF row.
+   * For fixed-length types, the alignment is equivalent to the type size in bytes.
+   *
+   * @param rapidsType a variable length data type.
+   * @return the alignment in bytes of the RAPIDS type in the variable width JCUDF.
+   */
+  public static int getDataAlignmentForDataType(DType rapidsType) {
+    switch (rapidsType.getTypeId()) {
+      case STRING:
+        // Strings are 1-byte aligned
+        return JCUDF_STRING_TYPE_ALIGNMENT;
+      case LIST:
+        // Data aligned depends on the datatype. return 8 bytes for now.
+      case STRUCT:
+        // Structs are 8-bytes aligned
+        return JCUDF_VAR_LENGTH_DATA_DEFAULT_ALIGNMENT;
+      default:
+        // all other types are aligned based on the size.
+        return rapidsType.getSizeInBytes();
+    }
+  }
+
+  /**
+   * Sets an estimate size in bytes for variable-length data types.
+   * This is used to get a rough estimate of the bytes needed to allocate JCUDF row.
+   *
+   * @param rapidsType a variable length data type.
+   * @return size in bytes of the variable length datatype.
+   */
+  public static int getEstimateSizeForVarLengthTypes(DType rapidsType) {
+    switch (rapidsType.getTypeId()) {
+      case STRING:
+        return JCUDF_TYPE_STRING_LENGTH_ESTIMATE;
+      case LIST:
+      case STRUCT:
+        // Return 32 bytes for other types until we decide on how to handle each type.
```

Comment:
Is there a missing return 32 here?

Reply:
Thanks @jbrennan333 
I should remove those and keep the default case since we do not support LIST and STRUCT yet. And fix the typo bytes vs. bits